### PR TITLE
Use sizeof more pervasively

### PR DESCRIPTION
**Pull request by compnerd**
Change a few places to use `sizeof` rather than hard coding the size of structures and types based on pointer size.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @MadCoder

File: src/queue_internal.h [hunk 1]
```diff
@@ -784,17 +782,16 @@ dispatch_queue_attr_t _dispatch_get_default_queue_attr(void);
...
 	void *dc_other
-#define _DISPATCH_SIZEOF_PTR 4
 #endif
 #define _DISPATCH_CONTINUATION_PTRS 8
 #if DISPATCH_HW_CONFIG_UP
 // UP devices don't contend on continuations so we don't need to force them to
 // occupy a whole cacheline (which is intended to avoid contention)
 #define DISPATCH_CONTINUATION_SIZE \
-		(_DISPATCH_CONTINUATION_PTRS * _DISPATCH_SIZEOF_PTR)
+		(_DISPATCH_CONTINUATION_PTRS * __SIZEOF_POINTER__)
```
Comment (MadCoder) [inline src/queue_internal.h]: `__SIZEOF_POINTER__` is not standard afaict which is why dispatch does it this way.

Comment (compnerd) [inline src/queue_internal.h]: Right, it's not standard.  Isn't clang a requisite though?  The other ones seem fine though, so, let me drop this part, and we can address it separately.

Comment (compnerd) [inline src/queue_internal.h]: BTW, is `__LP64__` standard?  I don't believe that MSVC defines that.

**Comment by compnerd** [comment]
Dropped the `__SIZEOF_POINTER__` changes, which we can discuss separately.

**Comment by compnerd** [comment]
@MadCoder anything else for this part of the change?

File: src/allocator_internal.h [hunk 2]
```diff
@@ -147,12 +147,6 @@ typedef unsigned long bitmap_t;
 
 #define PADDING_TO_CONTINUATION_SIZE(x) (ROUND_UP_TO_CONTINUATION_SIZE(x) - (x))
 
-#if defined(__LP64__)
-#define SIZEOF_HEADER 16
-#else
-#define SIZEOF_HEADER 8
-#endif
-
```
Comment (MadCoder) [inline src/allocator_internal.h]: I'd rather keep the SIZEOF_HEADER to a define to the right sizeof()

File: src/allocator_internal.h [hunk 3]
```diff
@@ -162,12 +156,12 @@ typedef unsigned long bitmap_t;
...
 // so that we can perform first_bitmap_in_same_page()
 #define SUPERMAPS_TO_MAPS_PADDING (PADDING_TO_BITMAP_ALIGNMENT_AND_CONTINUATION_SIZE( \
-		SIZEOF_SUPERMAPS + HEADER_TO_SUPERMAPS_PADDING + SIZEOF_HEADER))
+		SIZEOF_SUPERMAPS + HEADER_TO_SUPERMAPS_PADDING + sizeof(struct dispatch_magazine_header_s)))
 
 #define MAPS_TO_FPMAPS_PADDING (PADDING_TO_CONTINUATION_SIZE(SIZEOF_MAPS))
 
 #define BYTES_LEFT_IN_FIRST_PAGE (BYTES_PER_PAGE - \
-		(SIZEOF_HEADER + HEADER_TO_SUPERMAPS_PADDING + SIZEOF_SUPERMAPS + \
+		(sizeof(struct dispatch_magazine_header_s) + HEADER_TO_SUPERMAPS_PADDING + SIZEOF_SUPERMAPS + \
```
Comment (MadCoder) [inline src/allocator_internal.h]: and keep SIZEOF_HEADER here which has the side effect of keeping it to 80 columns

File: src/allocator_internal.h [hunk 4]
```diff
@@ -97,25 +97,20 @@
 // Use the largest type your platform is comfortable doing atomic ops with.
 // TODO: rdar://11477843
 typedef unsigned long bitmap_t;
-#if defined(__LP64__)
-#define BYTES_PER_BITMAP 8
```
Comment (MadCoder) [inline src/allocator_internal.h]: same as the other guy, makes this `sizeof(bitmap_t)`

File: src/allocator_internal.h [hunk 5]
```diff
@@ -97,25 +97,20 @@
...
-#define BYTES_PER_BITMAP 4
-#endif
 
 #define BITMAP_C(v) ((bitmap_t)(v))
 #define BITMAP_ALL_ONES (~BITMAP_C(0))
 
 // Stop configuring.
 
-#define CONTINUATIONS_PER_BITMAP (BYTES_PER_BITMAP * 8)
+#define CONTINUATIONS_PER_BITMAP (sizeof(bitmap_t) * 8)
```
Comment (MadCoder) [inline src/allocator_internal.h]: if you're feeling pedantic, that 8 should be CHAR_BIT

File: src/allocator_internal.h [hunk 6]
```diff
@@ -138,8 +133,8 @@ typedef unsigned long bitmap_t;
 // this will round up such that first_bitmap_in_same_page() can mask the address
 // of a bitmap_t in the maps to obtain the first bitmap for that same page
 #define ROUND_UP_TO_BITMAP_ALIGNMENT(x) \
-		(((x) + ((BITMAPS_PER_PAGE * BYTES_PER_BITMAP) - 1u)) & \
-		~((BITMAPS_PER_PAGE * BYTES_PER_BITMAP) - 1u))
```
Comment (MadCoder) [inline src/allocator_internal.h]: revert

File: src/allocator_internal.h [hunk 7]
```diff
@@ -148,7 +143,7 @@ typedef unsigned long bitmap_t;
 #define PADDING_TO_CONTINUATION_SIZE(x) (ROUND_UP_TO_CONTINUATION_SIZE(x) - (x))
 
 #define SIZEOF_SUPERMAPS (BYTES_PER_SUPERMAP * SUPERMAPS_PER_MAGAZINE)
-#define SIZEOF_MAPS (BYTES_PER_BITMAP * BITMAPS_PER_SUPERMAP * \
+#define SIZEOF_MAPS (sizeof(bitmap_t) * BITMAPS_PER_SUPERMAP * \
```
Comment (MadCoder) [inline src/allocator_internal.h]: revert

File: src/allocator_internal.h [hunk 8]
```diff
@@ -171,7 +166,7 @@ typedef unsigned long bitmap_t;
 #define REMAINDER_IN_FIRST_PAGE (BYTES_LEFT_IN_FIRST_PAGE - \
 		(FULL_BITMAPS_IN_FIRST_PAGE * CONSUMED_BYTES_PER_BITMAP) - \
 		(FULL_BITMAPS_IN_FIRST_PAGE ? 0 : \
-		ROUND_UP_TO_CONTINUATION_SIZE(BYTES_PER_BITMAP)))
```
Comment (MadCoder) [inline src/allocator_internal.h]: revert

File: src/allocator_internal.h [hunk 9]
```diff
@@ -181,7 +176,7 @@ typedef unsigned long bitmap_t;
 		(REMAINDERED_CONTINUATIONS_IN_FIRST_PAGE == 0 ? 0 : 1))
 
 #define FPMAPS_TO_FPCONTS_PADDING (PADDING_TO_CONTINUATION_SIZE(\
-		BYTES_PER_BITMAP * BITMAPS_IN_FIRST_PAGE))
```
Comment (MadCoder) [inline src/allocator_internal.h]: revert

File: src/allocator_internal.h [hunk 10]
```diff
@@ -162,12 +156,12 @@ typedef unsigned long bitmap_t;
 // we want to align the maps to a continuation size, but we must also have proper padding
 // so that we can perform first_bitmap_in_same_page()
 #define SUPERMAPS_TO_MAPS_PADDING (PADDING_TO_BITMAP_ALIGNMENT_AND_CONTINUATION_SIZE( \
-		SIZEOF_SUPERMAPS + HEADER_TO_SUPERMAPS_PADDING + SIZEOF_HEADER))
+		SIZEOF_SUPERMAPS + HEADER_TO_SUPERMAPS_PADDING + sizeof(struct dispatch_magazine_header_s)))
```
Comment (MadCoder) [inline src/allocator_internal.h]: revert

**Comment by compnerd** [comment]
Updated and also changed the `8` to `CHAR_BIT`.

**Comment by MadCoder** [comment]
@swift-ci please test

**Comment by MadCoder** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@MadCoder good to go on this?

**Comment by das** [comment]
I'm going to revert this, it breaks the build on Darwin with:

```
In file included from libdispatch/src/allocator.c:22:
libdispatch/src/allocator_internal.h:240:5: error: function-like macro 'sizeof' is not defined
#if SUPERMAPS_TO_MAPS_PADDING > 0
    ^
libdispatch/src/allocator_internal.h:168:3: note: expanded from macro 'SUPERMAPS_TO_MAPS_PADDING'
                SIZEOF_SUPERMAPS + HEADER_TO_SUPERMAPS_PADDING + SIZEOF_HEADER))
                ^
libdispatch/src/allocator_internal.h:159:27: note: expanded from macro 'SIZEOF_SUPERMAPS'
#define SIZEOF_SUPERMAPS (BYTES_PER_SUPERMAP * SUPERMAPS_PER_MAGAZINE)
                          ^
libdispatch/src/allocator_internal.h:125:28: note: expanded from macro 'BYTES_PER_SUPERMAP'
#define BYTES_PER_SUPERMAP sizeof(bitmap_t)
                           ^
libdispatch/src/allocator_internal.h:248:5: error: function-like macro 'sizeof' is not defined
#if MAPS_TO_FPMAPS_PADDING > 0
    ^
libdispatch/src/allocator_internal.h:170:62: note: expanded from macro 'MAPS_TO_FPMAPS_PADDING'
#define MAPS_TO_FPMAPS_PADDING (PADDING_TO_CONTINUATION_SIZE(SIZEOF_MAPS))
                                                             ^
libdispatch/src/allocator_internal.h:160:22: note: expanded from macro 'SIZEOF_MAPS'
#define SIZEOF_MAPS (BYTES_PER_BITMAP * BITMAPS_PER_SUPERMAP * \
                     ^
libdispatch/src/allocator_internal.h:111:26: note: expanded from macro 'BYTES_PER_BITMAP'
#define BYTES_PER_BITMAP sizeof(bitmap_t)
                         ^
libdispatch/src/allocator_internal.h:258:5: error: function-like macro 'sizeof' is not defined
#if FPMAPS_TO_FPCONTS_PADDING > 0
    ^
libdispatch/src/allocator_internal.h:193:3: note: expanded from macro 'FPMAPS_TO_FPCONTS_PADDING'
                BYTES_PER_BITMAP * BITMAPS_IN_FIRST_PAGE))
                ^
libdispatch/src/allocator_internal.h:111:26: note: expanded from macro 'BYTES_PER_BITMAP'
#define BYTES_PER_BITMAP sizeof(bitmap_t)
                         ^
libdispatch/src/allocator_internal.h:278:5: error: function-like macro 'sizeof' is not defined
#if AFTER_CONTS_PADDING > 0
    ^
libdispatch/src/allocator_internal.h:202:33: note: expanded from macro 'AFTER_CONTS_PADDING'
                (DISPATCH_CONTINUATION_SIZE * CONTINUATIONS_PER_MAGAZINE)))
                                              ^
libdispatch/src/allocator_internal.h:140:4: note: expanded from macro 'CONTINUATIONS_PER_MAGAZINE'
                (BITMAPS_PER_MAGAZINE * CONTINUATIONS_PER_BITMAP)
                 ^
libdispatch/src/allocator_internal.h:138:31: note: expanded from macro 'BITMAPS_PER_MAGAZINE'
#define BITMAPS_PER_MAGAZINE (SUPERMAPS_PER_MAGAZINE * BITMAPS_PER_SUPERMAP)
                              ^
libdispatch/src/allocator_internal.h:137:3: note: expanded from macro 'SUPERMAPS_PER_MAGAZINE'
                CONSUMED_BYTES_PER_SUPERMAP)
                ^
libdispatch/src/allocator_internal.h:126:38: note: expanded from macro 'CONSUMED_BYTES_PER_SUPERMAP'
#define CONSUMED_BYTES_PER_SUPERMAP (BYTES_PER_SUPERMAP + \
                                     ^
libdispatch/src/allocator_internal.h:125:28: note: expanded from macro 'BYTES_PER_SUPERMAP'
#define BYTES_PER_SUPERMAP sizeof(bitmap_t)
                           ^
```

**Pull request merged**